README.md
queue_manager [![Build Status](https://api.travis-ci.org/WPI-RAIL/queue_manager.png)](https://travis-ci.org/WPI-RAIL/queue_manager)
=============

#### Server Side ROS Queue Node
For full documentation, see [the ROS wiki](http://ros.org/wiki/rail_segmentation).

### License
queue_manager is released with a BSD license. For full terms and conditions, see the [LICENSE](LICENSE) file.

### Authors
See the [AUTHORS.md](AUTHORS.md) file for a full list of contributors.

### Notes From Peter
Topics:
 - rms_queue - publishes for each user: user id, time left in queue (valid for non 1st user only), and time left in study (valid for 1st user only)

Services:
 - update_queue - called by the web interface when someone: loads the webpage, reloads the webpage, exist the browser, or navigates to another website

### Known Bugs
 - study time can go negative when all users leave the queue. This has never caused any inappropriate behavior though, since it only happens when everyone leaves
 - study time described in the Ros_Queue object may not be respected, and will default to 10 minute trials. It should work but I has at times not worked. At least you know it's possible.


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package queue_manager
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.1 (2015-05-07)
------------------
* cleanup so it will compile
* Merge pull request #1 from PeterMitrano/develop
  moved to queue_manager
* updated to matched repo name, and added info in the readme
* moved to queue_manager
* Initial commit
* Contributors: Russell Toris, peter mitrano, petermitrano


src/queue_manager.cpp
//
// Created by pdmitrano on 4/1/15.
//

#include <queue_manager/queue_manager.h>

using namespace std;

int main(int argc, char *argv[])
{
  ros::init(argc, argv, "queue_manager");

  //construct a queue
  QueueManager queue_manager;

  return EXIT_SUCCESS;
}

QueueManager::QueueManager() : countdown_(DEFAULT_TRIAL), run_countdown_(false)
{
  //sent out queue
  ros::Publisher queue_pub = private_node_.advertise<queue_manager::Queue>("queue", 1000);
  //add user to queue when someone new visits the website
  ros::ServiceServer update_queue_serv = private_node_.advertiseService("update_queue",
                                                                        &QueueManager::onUpdateQueue, this);

  //loop at the rate of LOOP_RATE, and publish queue
  ros::Rate r(LOOP_RATE);

  ROS_INFO("publishing queue...");

  while (ros::ok())
  {

    ROS_INFO("countdown: %i\n", countdown_);

    //only run the countdown when the queue isn't empty
    if (run_countdown_)
    {
      //when you countdown has reached 0, reset it and remove the first/current user
      if (!countdown_)
      {
        queue_.pop_front(); //bye bye!
        //reset count down to next users trial time
        countdown_ = queue_.begin()->second;
      }

      //stop counting if the queue is now empty
      if (queue_.empty())
      {
        run_countdown_ = false;
      }

      countdown_--;
    }

    queue_manager::Queue queue_message;

    deque<pair<int, int> >::iterator it = queue_.begin();
    int position = 0;
    int wait_t;
    int t_left;

    //create queue to publish
    while (it != queue_.end())
    {
      pair<int, int> user = *(it++);

      queue_manager::UserStatus user_status;
      user_status.user_id = user.first;

      //calculate wait time in seconds

      if (position > 0)
      {
        wait_t = countdown_ + (position - 1) * user.second;
      }
      else if (position == 0)
      {
        wait_t = -1;
        t_left = countdown_;
      }

      ros::Duration wait_time(wait_t);
      ros::Duration time_left(t_left);
      user_status.wait_time = wait_time;
      user_status.time_left = time_left;
      queue_message.queue.push_back(user_status);

      position++;

      ROS_INFO("user %i", user_status.user_id);
    }

    //publish the queue message
    queue_pub.publish(queue_message);

    ros::spinOnce();
    r.sleep();
  }
}

bool QueueManager::onUpdateQueue(queue_manager::UpdateQueue::Request & req, queue_manager::UpdateQueue::Response & res)
{
  //get the user Id from the message
  int user_id = req.user_id;
  int study_time = req.study_time;

  //iterate over the queue, and erase the user that matches the user_id in the message
  deque<pair<int, int> >::iterator it = queue_.begin();
  while (it != queue_.end())
  {
    if (user_id == (*it).first)
    {
      if (!req.enqueue)
      {
        //when first user leaves reset time for the next user!
        if (it == queue_.begin())
        {
          queue_.pop_front();
          int next_study_time = queue_.front().second;
          countdown_ = next_study_time ? next_study_time : DEFAULT_TRIAL;
        } else
        {
          queue_.erase(it);
        }
      }
      return true;
    }
    it++;
  }

  if (req.enqueue)
  {
    study_time = study_time ? study_time : DEFAULT_TRIAL;

    //add that id to the back of the deque
    queue_.push_back(pair<int, int>(user_id, study_time));
    countdown_ = study_time;

    //start counting down
    run_countdown_ = true;
  }
  return true;
}


src/rms_queue_manager.cpp
//
// Created by pdmitrano on 4/1/15.
//

#include <queue_manager/queue_manager.h>

int main(int argc, char *argv[])
{
  ros::init(argc, argv, "queue_manager");

  //construct a queue
  RMS_Queue_Manager queue_manager;

  return EXIT_SUCCESS;
}

RMS_Queue_Manager::RMS_Queue_Manager()
    : countdown_(DEFAULT_TRIAL),
      run_countdown(false)
{

  ros::NodeHandle n;

  //sent out queue
  ros::Publisher queue_pub = n.advertise<queue_manager::RMSQueue>("rms_queue", 1000);

  //add user to queue when someone new visits the website
  ros::ServiceServer update_queue_serv = n.advertiseService("update_queue", &RMS_Queue_Manager::on_update_queue, this);

  //loop at the rate of LOOP_RATE, and publish queue
  ros::Rate r(RMS_Queue_Manager::LOOP_RATE);

  ROS_INFO("publishing queue...");

  while (ros::ok())
  {

    ROS_INFO("countdown: %i\n", countdown_);

    //only run the countdown when the queue isn't empty
    if (run_countdown)
    {
      //when you countdown has reached 0, reset it and remove the first/current user
      if (!countdown_)
      {
        queue_.pop_front(); //bye bye!
        //reset count down to next users trial time
        countdown_ = queue_.begin()->second;
      }

      //stop counting if the queue is now empty
      if (queue_.empty())
      {
        run_countdown = false;
      }

      countdown_--;
    }

    queue_manager::RMSQueue queue_message;

    std::deque<std::pair<int, int> >::iterator it = queue_.begin();
    int position = 0;
    int wait_t;
    int t_left;

    //create queue to publish
    while (it != queue_.end())
    {
      std::pair<int, int> user = *(it++);

      queue_manager::UserStatus user_status;
      user_status.user_id = user.first;

      //calculate wait time in seconds

      if (position > 0)
      {
        wait_t = countdown_ + (position - 1) * user.second;
      }
      else if (position == 0)
      {
        wait_t = -1;
        t_left = countdown_;
      }

      ros::Duration wait_time(wait_t);
      ros::Duration time_left(t_left);
      user_status.wait_time = wait_time;
      user_status.time_left = time_left;
      queue_message.queue.push_back(user_status);

      position++;

      ROS_INFO("user %i", user_status.user_id);
    }

    //publish the queue message
    queue_pub.publish(queue_message);

    ros::spinOnce();
    r.sleep();
  }
}

bool RMS_Queue_Manager::on_update_queue(queue_manager::UpdateQueue::Request &req,
                                        queue_manager::UpdateQueue::Response &res)
{
  //get the user Id from the message
  int user_id = req.user_id;
  int study_time = req.study_time;

  //iterate over the queue, and erase the user that matches the user_id in the message
  std::deque<std::pair<int, int> >::iterator it = queue_.begin();
  while (it != queue_.end())
  {
    if (user_id == (*it).first)
    {
      if (!req.enqueue)
      {
        //when first user leaves reset time for the next user!
        if (it == queue_.begin())
        {
          queue_.pop_front();
          int next_study_time = queue_.front().second;
          countdown_ = next_study_time ? next_study_time : RMS_Queue_Manager::DEFAULT_TRIAL;
        } else
        {
          queue_.erase(it);
        }
      }
      return true;
    }
    it++;
  }

  if (req.enqueue)
  {

    study_time = study_time ? study_time : RMS_Queue_Manager::DEFAULT_TRIAL;

    //add that id to the back of the deque
    queue_.push_back(std::pair<int, int>(user_id, study_time));
    countdown_ = study_time;

    //start counting down
    run_countdown = true;
  }
  return true;
}


include/queue_manager/queue_manager.h
//
// Created by pdmitrano on 4/1/15.
//

#ifndef QUEUE_MANAGER_H_
#define QUEUE_MANAGER_H_

#include <ros/ros.h>
#include <queue_manager/Queue.h>
#include <queue_manager/UserStatus.h>
#include <queue_manager/UpdateQueue.h>
#include <deque>

class QueueManager
{
public:
  /*!brief The frequency of the loop in (1/seconds). This is how often we publish the queue */
  static const int LOOP_RATE = 1;
  /*! default length of trial is 600 seconds. This will always be overwritten if the enqueue */
  static const int DEFAULT_TRIAL = 600;

  /*!
   * \brief This starts the server, waits for param server and create clickable parking spots
   */
  QueueManager();

private:
  /*! the queue to hold the user_ids and wait times in order. */
  std::deque<std::pair<int, int> > queue_;
  /*! keeps track of the time left for the active user */
  int countdown_;
  /*! run countdown if there is someone in the queue */
  bool run_countdown_;
  /*! Private node handle. */
  ros::NodeHandle private_node_;

  /*!
   * \brief Add or remove a user to the deque
   */
  bool onUpdateQueue(queue_manager::UpdateQueue::Request &req, queue_manager::UpdateQueue::Response &res);

};

#endif


include/rms_queue_manager/rms_queue_manager.h
//
// Created by pdmitrano on 4/1/15.
//

#ifndef _PROJECT_RMS_QUEUE_MANAGER_H_
#define _PROJECT_RMS_QUEUE_MANAGER_H_

#include "ros/ros.h"
#include <queue_manager/RMSQueue.h> //custom message file with array of UserStatus
#include <queue_manager/UserStatus.h> //custom message file with user_id and wait_time
#include <queue_manager/UpdateQueue.h>
#include "std_msgs/Int32MultiArray.h"
#include "std_msgs/Int32.h"
#include <deque>

class RMS_Queue_Manager
{
public:
  /**
  * \brief The frequency of the loop in (1/seconds). This is how often we publish the queue
  */
  static const int LOOP_RATE = 1;

  /**
  * \brief default length of trial is 600 seconds. This will always be overwritten if the enqueue
  */
  static const int DEFAULT_TRIAL = 600;

  /**
  * \brief Constructor
  *	This starts the server, waits for param server and create clickable parking spots
  */
  RMS_Queue_Manager();

private:

  /**
  * the queue to hold the user_ids and wait times in order.
  */
  std::deque<std::pair<int, int> > queue_;

  /**
  * keeps track of the time left for the active user
  */
  int countdown_;

  /**
  * run countdown if there is someone in the queue
  */
  bool run_countdown;

  /**
  * \brief Add or remove a user to the deque
  */
  bool on_update_queue(queue_manager::UpdateQueue::Request &req,
                       queue_manager::UpdateQueue::Response &res);

};

#endif //_PROJECT_RMS_QUEUE_MANAGER_H_


src/rail_user_queue_manager.cpp
//
// Created by pdmitrano on 4/1/15.
//

#include <rail_user_queue_manager/rail_user_queue_manager.h>

using namespace std;
using namespace rail;

int main(int argc, char *argv[])
{
  ros::init(argc, argv, "queue_manager");

  //construct a queue
  UserQueueManager queue_manager;

  return EXIT_SUCCESS;
}

UserQueueManager::UserQueueManager() : countdown_(DEFAULT_TRIAL), run_countdown_(false), private_node_("~")
{
  //sent out queue
  ros::Publisher queue_pub = private_node_.advertise<rail_user_queue_manager::Queue>("queue", 1000);
  //add user to queue when someone new visits the website
  ros::ServiceServer update_queue_serv = private_node_.advertiseService("update_queue",
                                                                        &UserQueueManager::onUpdateQueue, this);

  //loop at the rate of LOOP_RATE, and publish queue
  ros::Rate r(LOOP_RATE);

  ROS_INFO("publishing queue...");

  while (ros::ok())
  {
    // ROS_INFO("countdown: %i\n", countdown_);

    //only run the countdown when the queue isn't empty
    if (run_countdown_)
    {
      //when you countdown has reached 0, reset it and remove the first/current user
      if (!countdown_)
      {
        queue_.pop_front(); //bye bye!
        //reset count down to next users trial time
        countdown_ = queue_.begin()->second;
      }

      //stop counting if the queue is now empty
      if (queue_.empty())
      {
        run_countdown_ = false;
      }

      countdown_--;
    }

    rail_user_queue_manager::Queue queue_message;

    deque<pair<int, int> >::iterator it = queue_.begin();
    int position = 0;
    int wait_t;
    int t_left;

    //create queue to publish
    while (it != queue_.end())
    {
      pair<int, int> user = *(it++);

      rail_user_queue_manager::UserStatus user_status;
      user_status.user_id = user.first;

      //calculate wait time in seconds

      if (position > 0)
      {
        wait_t = countdown_ + (position - 1) * user.second;
      }
      else if (position == 0)
      {
        wait_t = -1;
        t_left = countdown_;
      }

      ros::Duration wait_time(wait_t);
      ros::Duration time_left(t_left);
      user_status.wait_time = wait_time;
      user_status.time_left = time_left;
      queue_message.queue.push_back(user_status);

      position++;

      ROS_INFO("user %i", user_status.user_id);
    }

    //publish the queue message
    queue_pub.publish(queue_message);

    ros::spinOnce();
    r.sleep();
  }
}

bool UserQueueManager::onUpdateQueue(rail_user_queue_manager::UpdateQueue::Request &req,
                                     rail_user_queue_manager::UpdateQueue::Response &res)
{
  //get the user Id from the message
  int user_id = req.user_id;
  int study_time = req.study_time;

  //iterate over the queue, and erase the user that matches the user_id in the message
  deque<pair<int, int> >::iterator it = queue_.begin();
  while (it != queue_.end())
  {
    if (user_id == (*it).first)
    {
      if (!req.enqueue)
      {
        //when first user leaves reset time for the next user!
        if (it == queue_.begin())
        {
          queue_.pop_front();
          int next_study_time = queue_.front().second;
          countdown_ = next_study_time ? next_study_time : DEFAULT_TRIAL;
        } else
        {
          queue_.erase(it);
        }
      }
      return true;
    }
    it++;
  }

  if (req.enqueue)
  {
    study_time = study_time ? study_time : DEFAULT_TRIAL;

    //add that id to the back of the deque
    queue_.push_back(pair<int, int>(user_id, study_time));
    countdown_ = study_time;

    //start counting down
    run_countdown_ = true;
  }
  return true;
}
